qcodes/instrument/parameter.py

```diff
@@ -94,83 +94,89 @@ class _BaseParameter(Metadatable, DeferredOperations):
             ``update=True``, for example if it takes too long to update.
             Default True.
 
+        max_val_age (Optional[int]): The max time (in seconds) to trust a
+            saved value obtained from get_latest(). If this parameter has not
+            been set or measured more recently than this, perform an
+            additional measurement.
+
         metadata (Optional[dict]): extra information to include with the
             JSON snapshot of the parameter
     """
     def __init__(self, name, instrument, snapshot_get, metadata,
-                 snapshot_value=True):
+                 step=None, scale=None, inter_delay=0, post_delay=0,
+                 val_mapping=None, get_parser=None, set_parser=None,
+                 snapshot_value=True, max_val_age=None, vals=None):
         super().__init__(metadata)
-        self._snapshot_get = snapshot_get
         self.name = str(name)
         self._instrument = instrument
+        self._snapshot_get = snapshot_get
         self._snapshot_value = snapshot_value
 
-        self.has_get = hasattr(self, 'get')
-        self.has_set = hasattr(self, 'set')
+        if not isinstance(vals, (Validator, type(None))):
+            raise TypeError('vals must be None or a Validator')
+        self.vals = vals
+
+        self.step = step
+        self.scale = scale
+        self.raw_value = None
+        self.inter_delay = inter_delay
+        self.post_delay = post_delay
 
-        if not (self.has_get or self.has_set):
-            raise AttributeError('A parameter must have either a get or a '
-                                 'set method, or both.')
+        # TODO (nulinspiratie) handle int vs string conversion in val_mapping
```

Comment:
Yeah this is the one last thing I need to implement, I'll do it soon. 

The issue is that a val mapping can map either strings or ints. For instance, a message returned by an instrument could always be a string, even though the string would be either `'0'` or `'1'`. The old val_mapping used to attempt to convert the value into a string and then also check if it had a mapping.